### PR TITLE
SPARKC-359: support client certificate authentication for two-way SSL encryption

### PR DESCRIPTION
**Pull request by bchen-talend**

### CODE REVIEW

**Comment by RussellSpitzer** [comment]
Can you please make a corresponding Jira for this https://datastax-oss.atlassian.net/projects/SPARKC/issues . It makes it a lot easier for us to track. This is a great addition

**Comment by bchen-talend** [comment]
Thanks, here is the [jira](https://datastax-oss.atlassian.net/browse/SPARKC-359)

**Comment by RussellSpitzer** [comment]
This looks good to me @jacek-lewandowski can you take a quick look at this?

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/CassandraConnectionFactory.scala [hunk 1]
```diff
@@ -77,9 +77,33 @@ object DefaultConnectionFactory extends CassandraConnectionFactory {
         } finally {
           IOUtils.closeQuietly(trustStoreFile)
         }
-
+        val kmf = if(conf.clientAuthEnabled){
+            conf.keyStorePath map {
+              case path ⇒
```
Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/CassandraConnectionFactory.scala]: please don't use this special characters, you can achieve nice looking arrows by ligatures

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/CassandraConnectorConf.scala [hunk 2]
```diff
@@ -46,7 +46,11 @@ object CassandraConnectorConf extends Logging {
     trustStorePassword: Option[String] = None,
     trustStoreType: String = "JKS",
     protocol: String = "TLS",
-    enabledAlgorithms: Set[String] = Set("TLS_RSA_WITH_AES_128_CBC_SHA", "TLS_RSA_WITH_AES_256_CBC_SHA")
+    enabledAlgorithms: Set[String] = Set("TLS_RSA_WITH_AES_128_CBC_SHA", "TLS_RSA_WITH_AES_256_CBC_SHA"),
+    clientAuthEnabled: Boolean = false,
```
Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/CassandraConnectorConf.scala]: Do we really need that boolean parameter? Given we have`keyStorePath` optional, it `clientAuthEnabled` seems redundant.

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/CassandraConnectionFactory.scala [hunk 3]
```diff
@@ -77,9 +77,33 @@ object DefaultConnectionFactory extends CassandraConnectionFactory {
         } finally {
           IOUtils.closeQuietly(trustStoreFile)
         }
-
+        val kmf = if(conf.clientAuthEnabled){
```
Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/CassandraConnectionFactory.scala]: nit: formatting

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/CassandraConnectionFactory.scala [hunk 4]
```diff
@@ -77,9 +77,33 @@ object DefaultConnectionFactory extends CassandraConnectionFactory {
...
+            conf.keyStorePath map {
+              case path ⇒
+                val keyStoreFile = new FileInputStream(path)
+                val kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm)
+                val keyStore = KeyStore.getInstance(conf.keyStoreType)
+                conf.keyStorePassword match {
```
Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/CassandraConnectionFactory.scala]: Looks like you repeating the same code twice, while only password should be mapped:
```scala
val keyStorePassword = conf.keyStorePassword.map(_.toCharArray)
keyStore.load...
kmf.init...
```

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/CassandraConnectionFactory.scala [hunk 5]
```diff
@@ -77,9 +77,33 @@ object DefaultConnectionFactory extends CassandraConnectionFactory {
         } finally {
           IOUtils.closeQuietly(trustStoreFile)
         }
-
+        val kmf = if(conf.clientAuthEnabled){
+            conf.keyStorePath map {
+              case path ⇒
+                val keyStoreFile = new FileInputStream(path)
```
Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/CassandraConnectionFactory.scala]: The stream is not closed

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/CassandraConnectionFactory.scala [hunk 6]
```diff
@@ -77,9 +77,33 @@ object DefaultConnectionFactory extends CassandraConnectionFactory {
...
+                  }
+                }
+                kmf
+            }
+          }else{
+            None
+          }
         val context = SSLContext.getInstance(conf.protocol)
-        context.init(null, tmf.getTrustManagers, new SecureRandom)
+        kmf match {
```
Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/CassandraConnectionFactory.scala]: We should either support combination when keystore is specified but truststore is not or throw exception in such a case.

**Comment by jacek-lewandowski** [comment]
please update CHANGES.txt

**Comment by jacek-lewandowski** [comment]
@RussellSpitzer i think we need to change the mechanism of configuring Cassandra with template yamls one yaml per configuration...

**Comment by RussellSpitzer** [comment]
:| I really don't want to have more custom yaml configuration stuff but you are right

**Comment by ndamclean** [comment]
Is anyone planning on cleaning this up / merging it? I would like to be able to use this feature.

**Comment by jacek-lewandowski** [comment]
I can clean this up

**Comment by indigodp** [comment]
Hi, will this request be merged and released soon? It is very important feature.

**Comment by mparkachov** [comment]
Hi, I'm very interested in this request, or in other words, I need this feature badly :)

**Comment by jacek-lewandowski** [comment]
As you can see, this https://datastax-oss.atlassian.net/browse/SPARKC-359 is already solved.

**Pull request closed without merging**